File: backend/src/agent/ipalloc.zig
//! Sequential guest-IP allocation from a CIDR.
//!
//! The CIDR's `.1` host is the bridge gateway; guests are handed out
//! sequentially starting at `.2`. Allocation is index-based and stable: a guest
//! assigned slot N always gets the same IP, so a restored/forked VM that records
//! its slot keeps its address. Freed slots are reused lowest-first.
const std = @import("std");

pub const Cidr = struct {
    /// Network base address in host byte order (e.g. 10.231.0.0 -> 0x0AE70000).
    base: u32,
    prefix: u6,

    /// Parse "a.b.c.d/n".
    pub fn parse(s: []const u8) ?Cidr {
        const slash = std.mem.indexOfScalar(u8, s, '/') orelse return null;
        const ip = parseIp4(s[0..slash]) orelse return null;
        const prefix = std.fmt.parseInt(u6, s[slash + 1 ..], 10) catch return null;
        if (prefix > 32) return null;
        return .{ .base = ip, .prefix = prefix };
    }

    /// Host count usable for guests = 2^(32-prefix) minus network/gw/broadcast.
    pub fn hostCount(self: Cidr) u32 {
        const host_bits: u5 = @intCast(32 - @as(u32, self.prefix));
        const total = @as(u64, 1) << host_bits;
        if (total <= 3) return 0; // network + gw(.1) + broadcast
        return @intCast(total - 3);
    }

    /// The gateway address (network .1) in host byte order.
    pub fn gateway(self: Cidr) u32 {
        return netBase(self) + 1;
    }

    /// Netmask in host byte order.
    pub fn mask(self: Cidr) u32 {
        if (self.prefix == 0) return 0;
        return ~@as(u32, 0) << @intCast(32 - @as(u32, self.prefix));
    }

    fn netBase(self: Cidr) u32 {
        return self.base & self.mask();
    }

    /// IP (host byte order) for guest slot `idx` (0-based), starting at .2.
    pub fn hostAt(self: Cidr, idx: u32) u32 {
        return netBase(self) + 2 + idx;
    }
};

/// Format a host-byte-order u32 as a dotted quad into `out`.
pub fn fmtIp(v: u32, out: []u8) []const u8 {
    return std.fmt.bufPrint(out, "{d}.{d}.{d}.{d}", .{
        (v >> 24) & 0xff, (v >> 16) & 0xff, (v >> 8) & 0xff, v & 0xff,
    }) catch out[0..0];
}

fn parseIp4(s: []const u8) ?u32 {
    var out: u32 = 0;
    var it = std.mem.splitScalar(u8, s, '.');
    var i: usize = 0;
    while (it.next()) |part| : (i += 1) {
        if (i >= 4) return null;
        const b = std.fmt.parseInt(u8, part, 10) catch return null;
        out = (out << 8) | b;
    }
    if (i != 4) return null;
    return out;
}

/// Tracks which guest slots are in use within a CIDR. `tap_<idx>` and the IP
/// for a VM both derive from its slot index. Caller serializes access.
pub const Allocator = struct {
    cidr: Cidr,
    used: std.DynamicBitSetUnmanaged,
    gpa: std.mem.Allocator,

    pub fn init(gpa: std.mem.Allocator, cidr: Cidr) !Allocator {
        const n = cidr.hostCount();
        var bits = try std.DynamicBitSetUnmanaged.initEmpty(gpa, n);
        errdefer bits.deinit(gpa);
        return .{ .cidr = cidr, .used = bits, .gpa = gpa };
    }

    pub fn deinit(self: *Allocator) void {
        self.used.deinit(self.gpa);
    }

    /// Claim the lowest free slot, returning its index, or null if exhausted.
    pub fn claim(self: *Allocator) ?u32 {
        var i: usize = 0;
        while (i < self.used.bit_length) : (i += 1) {
            if (!self.used.isSet(i)) {
                self.used.set(i);
                return @intCast(i);
            }
        }
        return null;
    }

    /// Reserve a specific slot (used when reconciling persisted instances).
    pub fn reserve(self: *Allocator, idx: u32) void {
        if (idx < self.used.bit_length) self.used.set(idx);
    }

    pub fn free(self: *Allocator, idx: u32) void {
        if (idx < self.used.bit_length) self.used.unset(idx);
    }

    pub fn ipFor(self: *const Allocator, idx: u32, out: []u8) []const u8 {
        return fmtIp(self.cidr.hostAt(idx), out);
    }
};

test "cidr parse and host addressing" {
    const c = Cidr.parse("10.231.0.0/24").?;
    try std.testing.expectEqual(@as(u6, 24), c.prefix);
    var buf: [16]u8 = undefined;
    try std.testing.expectEqualStrings("10.231.0.1", fmtIp(c.gateway(), &buf));
    try std.testing.expectEqualStrings("10.231.0.2", fmtIp(c.hostAt(0), &buf));
    try std.testing.expectEqualStrings("10.231.0.12", fmtIp(c.hostAt(10), &buf));
    try std.testing.expectEqualStrings("255.255.255.0", fmtIp(c.mask(), &buf));
    try std.testing.expectEqual(@as(u32, 253), c.hostCount());
}

test "allocator hands out sequential slots and reuses freed" {
    var alloc = try Allocator.init(std.testing.allocator, Cidr.parse("10.231.0.0/24").?);
    defer alloc.deinit();
    try std.testing.expectEqual(@as(u32, 0), alloc.claim().?);
    try std.testing.expectEqual(@as(u32, 1), alloc.claim().?);
    try std.testing.expectEqual(@as(u32, 2), alloc.claim().?);
    alloc.free(1);
    try std.testing.expectEqual(@as(u32, 1), alloc.claim().?); // reuse lowest free

    var buf: [16]u8 = undefined;
    try std.testing.expectEqualStrings("10.231.0.2", alloc.ipFor(0, &buf));
}

test "reserve marks slot used" {
    var alloc = try Allocator.init(std.testing.allocator, Cidr.parse("10.231.0.0/29").?);
    defer alloc.deinit();
    // /29 -> 8 addrs, minus net/gw/bcast = 5 host slots.
    try std.testing.expectEqual(@as(u32, 5), alloc.cidr.hostCount());
    alloc.reserve(0);
    alloc.reserve(2);
    try std.testing.expectEqual(@as(u32, 1), alloc.claim().?);
    try std.testing.expectEqual(@as(u32, 3), alloc.claim().?);
}

File: rust/agent/src/ipalloc.rs
//! Sequential guest-IP allocation from a CIDR.
//!
//! Port of backend/src/agent/ipalloc.zig — slot↔IP mapping, .2+slot, lowest-free.
//! The CIDR's `.1` host is the bridge gateway; guests start at `.2`.


#[derive(Debug, Clone, Copy)]
pub struct Cidr {
    /// Network base address in host byte order (e.g. 10.231.0.0 → 0x0AE70000).
    pub base: u32,
    pub prefix: u8,
}

impl Cidr {
    /// Parse "a.b.c.d/n".
    pub fn parse(s: &str) -> Option<Cidr> {
        let slash = s.find('/')?;
        let ip = parse_ip4(&s[..slash])?;
        let prefix: u8 = s[slash + 1..].parse().ok()?;
        if prefix > 32 { return None; }
        Some(Cidr { base: ip, prefix })
    }

    pub fn mask(self) -> u32 {
        if self.prefix == 0 { return 0; }
        !0u32 << (32 - self.prefix as u32)
    }

    pub fn net_base(self) -> u32 {
        self.base & self.mask()
    }

    /// Gateway address (.1) in host byte order.
    pub fn gateway(self) -> u32 {
        self.net_base() + 1
    }

    /// IP (host byte order) for guest slot `idx` (0-based), starting at .2.
    pub fn host_at(self, idx: u32) -> u32 {
        self.net_base() + 2 + idx
    }

    /// Host count usable for guests = 2^(32-prefix) minus network/gw/broadcast.
    pub fn host_count(self) -> u32 {
        let host_bits = 32u32.saturating_sub(self.prefix as u32);
        let total: u64 = 1u64 << host_bits;
        if total <= 3 { return 0; }
        (total - 3) as u32
    }
}

pub fn fmt_ip(v: u32) -> String {
    format!(
        "{}.{}.{}.{}",
        (v >> 24) & 0xff,
        (v >> 16) & 0xff,
        (v >> 8) & 0xff,
        v & 0xff
    )
}

fn parse_ip4(s: &str) -> Option<u32> {
    let parts: Vec<&str> = s.split('.').collect();
    if parts.len() != 4 { return None; }
    let mut out: u32 = 0;
    for part in &parts {
        let b: u8 = part.parse().ok()?;
        out = (out << 8) | b as u32;
    }
    Some(out)
}

/// Tracks which guest slots are in use within a CIDR.
pub struct Allocator {
    pub cidr: Cidr,
    used: Vec<bool>,
}

impl Allocator {
    pub fn new(cidr: Cidr) -> Self {
        let n = cidr.host_count() as usize;
        Allocator { cidr, used: vec![false; n] }
    }

    /// Claim the lowest free slot, returning its index, or None if exhausted.
    pub fn claim(&mut self) -> Option<u32> {
        for (i, used) in self.used.iter_mut().enumerate() {
            if !*used {
                *used = true;
                return Some(i as u32);
            }
        }
        None
    }

    /// Reserve a specific slot (used when reconciling persisted instances).
    pub fn reserve(&mut self, idx: u32) {
        let i = idx as usize;
        if i < self.used.len() {
            self.used[i] = true;
        }
    }

    pub fn free(&mut self, idx: u32) {
        let i = idx as usize;
        if i < self.used.len() {
            self.used[i] = false;
        }
    }

    pub fn ip_for(&self, idx: u32) -> String {
        fmt_ip(self.cidr.host_at(idx))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cidr_parse_and_host_addressing() {
        let c = Cidr::parse("10.231.0.0/24").unwrap();
        assert_eq!(c.prefix, 24);
        assert_eq!(fmt_ip(c.gateway()), "10.231.0.1");
        assert_eq!(fmt_ip(c.host_at(0)), "10.231.0.2");
        assert_eq!(fmt_ip(c.host_at(10)), "10.231.0.12");
        assert_eq!(fmt_ip(c.mask()), "255.255.255.0");
        assert_eq!(c.host_count(), 253);
    }

    #[test]
    fn test_slot_ip_math_10_231_0_0_24() {
        let c = Cidr::parse("10.231.0.0/24").unwrap();
        // slot 3 → .5 (net=.0, gw=.1, slot0=.2, slot1=.3, slot2=.4, slot3=.5)
        assert_eq!(fmt_ip(c.host_at(3)), "10.231.0.5");
        // The spec example: slot=3 → ip="10.231.0.3" in the meta fixture — that means
        // the fixture was written with slot=1 giving .3 which is 2+1=3. Let's verify:
        // slot=1 → net_base()+2+1 = 10.231.0.0+2+1 = 10.231.0.3
        assert_eq!(fmt_ip(c.host_at(1)), "10.231.0.3");
    }

    #[test]
    fn test_allocator_sequential_and_reuse() {
        let mut alloc = Allocator::new(Cidr::parse("10.231.0.0/24").unwrap());
        assert_eq!(alloc.claim(), Some(0));
        assert_eq!(alloc.claim(), Some(1));
        assert_eq!(alloc.claim(), Some(2));
        alloc.free(1);
        assert_eq!(alloc.claim(), Some(1)); // reuse lowest free
        assert_eq!(alloc.ip_for(0), "10.231.0.2");
    }

    #[test]
    fn test_reserve_marks_slot_used() {
        let mut alloc = Allocator::new(Cidr::parse("10.231.0.0/29").unwrap());
        // /29 -> 8 addrs, minus net/gw/bcast = 5 host slots.
        assert_eq!(alloc.cidr.host_count(), 5);
        alloc.reserve(0);
        alloc.reserve(2);
        assert_eq!(alloc.claim(), Some(1));
        assert_eq!(alloc.claim(), Some(3));
    }
}
